### ISSUE TITLE
Poor performance of diagonal transpose * vector

### ISSUE THREAD

**Issue by lruthotto** [JuliaLang/julia#21286]
Under the most current julia version, try this

```
v = randn(10000);
D = Diagonal(v);
@time D*v;
@time D'*v;
```
this gives me the following output:
```
0.000033 seconds (6 allocations: 78.359 KiB)
0.957613 seconds (6 allocations: 78.359 KiB)
```
seems like there is no tailored implementation for transpose*vector for Diagonal?

Even worse is the performance when multiplying with sparse matrix. This takes forever

```
julia> V = sprandn(10000,10000,0.01);
julia> D'*V
```

Here is my version info

```
julia> versioninfo()
Julia Version 0.6.0-pre.beta.30
Commit a5b993c* (2017-04-05 04:25 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```

**Comment by andreasnoack** [comment]
I fall back to the `Ac_mul_B!` generic implementation in both cases. This could fixes short term by adding a few of these methods for `Diagonal` but it will be solved more generally with the `Transpose` type for matrices.

**Comment by ararslan** [comment]
Seems like `Diagonal`-specific methods would be useful anyway, since `D'v == D*v == diag(D) .* v`

**Comment by fredrikekre** [comment]
Here are some other faulty things with `Diagonal`/`SparseMatrixCSC` I found looking at https://github.com/JuliaLang/julia/issues/20367 but then forgot about:

```julia
D = Diagonal(ones(4))
S = sprand(4, 4, 0.4)

D\S  # returns a "full" SparseMatrixCSC

S'*D # errors (works if eltype(D) = Complex)
D*S' # errors (works if eltype(D) = Complex)

S*D' # returns an Array
D'*S # interestingly enough this works
```

**Comment by georgemarrows** [comment]
Just noticed that PR #21302 does actually help with D' * sparse too, which the original bug report also mentioned. It's much faster (nearly 1000x) now, but allocates a lot more too, for reasons I haven't yet looked into. The result in both cases is a SparseMatrixCSC as you might hope.

```julia
D = Diagonal(randn(1000))
S = sprandn(1000,1000,0.01)
```

Master just before #21302
```
julia> gc(); @time (D' * S; nothing)
 15.771836 seconds (65 allocations: 523.203 KiB)

julia> gc(); @time (D' * S; nothing)
 15.820528 seconds (52 allocations: 522.453 KiB)
```

Master with #21302
```
julia> gc(); @time (D' * S; nothing)
  0.001998 seconds (37 allocations: 15.297 MiB)

julia> gc(); @time (D' * S; nothing)
  0.001965 seconds (37 allocations: 15.297 MiB)

julia> gc(); @time (D' * S; nothing)
  0.001947 seconds (37 allocations: 15.297 MiB)
````

**Comment by georgemarrows** [comment]
I think this issue can be closed:
- all the performance issues listed here (and one or two more) have been fixed by #21302
- I reported #21435 for @fredrikekre's comment above
